tests/sweep_runner/test_param_sweep_runner.py

```diff
@@ -455,11 +583,12 @@ def sweep_method(data: np.ndarray, param_1: int):  # type: ignore
 
     NO_OF_SWEEPS = 4
     EXPECTED_BLOCK_SHAPE = (
-        PREVIEWED_SLICES_SHAPE[0],
+        GLOBAL_SHAPE[0],
         NO_OF_SWEEPS,
-        PREVIEWED_SLICES_SHAPE[2],
+        GLOBAL_SHAPE[2],
     )
     assert runner.block.data.shape == EXPECTED_BLOCK_SHAPE
+    assert runner._vertical_slices_preview == 7
```

Comment:
I'm not particularly fond of this assertion, because it's checking something private in the sweep runner.

I also don't think it makes sense to fix the issue by just making that attribute public, because it's an internal value that the sweep runner is calculating to deal with paganin, I don't think it's something that should be exposed.

Stepping back, what we care about as far as using the sweep runner object is concerned, is that it updates the preview of the loader wrapper correctly to account for the paganin filter's kernel size. The loader wrapper's preview is public, and is what we ultimately care about, not the internal `ParamSweepRunner._vertical_slices_preview` value.

I also noticed that if I print out `loader.preview` before and after the sweep runner execution in the test, doing this instead at line 582:
```python
print(f"loader.preview before modifying: {loader.preview}")
runner.execute()
print(f"loader.preview after modifying: {loader.preview}")
```

then the preview in the `detector_y` dimension is seemingly incorrect; the start and stop are invalid (`start=48`, while `stop=3`):
```
loader.preview before execution: PreviewConfig(angles=PreviewDimConfig(start=0, stop=180),
detector_y=PreviewDimConfig(start=50, stop=53), detector_x=PreviewDimConfig(start=0, stop=160))
loader.preview after execution: PreviewConfig(angles=PreviewDimConfig(start=0, stop=180),
detector_y=PreviewDimConfig(start=48, stop=3), detector_x=PreviewDimConfig(start=0, stop=160))
```

So, while the private `self._vertical_slices_preview` value is what is expected (the test as it currently is, passes), the final public preview value looks like it's incorrect, and it's not being caught in the test.

I could be doing something wrong, so please do double-check if the preview is wrong in this test when you run it locally!

If it turns out that the preview is indeed incorrect, I think it would be better to remove the assertion on the private `self._vertical_slices_preview`, and instead assert that `loader.preview` is what we expect it to be, which takes care of both:
- the desire to assert only on public members of objects in tests
- asserting the value that ultimately we care about is correct (the updated preview)

Reply:
thanks @yousefmoazzam, Yes, good point, I've modified the tests accordingly. On the wrong preview stuff. I think there is an issue because of the `GLOBAL_SHAPE = (180, 3, 160)`, but we would like to extend the `preview` to be larger (7 slices, as we set `self._vertical_slices_preview = 7`). I guess, there is no check if the global size of data is actually smaller vertically than the `preview`. So I'm thinking that we should add it. I guess 3 vertical slices is the minimal size, so we can reset the `preview` to be something between 3 and 7, if the global size of data is smaller than 7?

Reply:
that becomes a bit fiddly may be with the indices... one possible way is to make `self._vertical_slices_preview = 5` and reject the previewed data that is smaller than that?  I think this is reasonable. Other methods should work on 5 slices well enough?

Reply:
OK, I've fixed that issue above by adding an additional condition on the size of the global data [here](https://github.com/DiamondLightSource/httomo/blob/57b60c0d8ae273e431b3063aea2d706c0e36b208/httomo/sweep_runner/param_sweep_runner.py#L142). I also added a test when the original (raw) data is too small for the sweep to work. It is set to 5 slices, which is reasonable. If the data is larger than this amount then we can manipulate with the preview: shrink or crop, but 5 is the condition.

Reply:
Ok, I admittedly don't really understand why 7 wasn't good and 5 is ok, but it feels like this detail may not be that significant in the bigger picture - the private member assertions have been replaced with a relevant public member assertion, and the test passes with the changes, so this feels much nicer now.

Reply:
no difference really, I just thought that 5 is probably enough for 3D processing, but it is so far purely heuristic tbh. We will see if that needs increasing with time. A good example will be iterative denoising in the sweep pipeline. This is where it becomes more obvious how many slices is sufficient enough to generalise it to a full 3D volume